### ISSUE TITLE
SEPA Mandates: this is incompatible with sql_mode=only_full_group_by

### ISSUE THREAD

**Issue by masetto** [Project60/org.project60.sepa#584]
Hello, when I click in "SEPA Mandates" tab in a contact I have this error:
```
 SELECT list is not in GROUP BY clause and contains nonaggregated column 'last.receive_date' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by]"
```
I'm using MySQL 5.7.3. Is it a prerequisite to remove `only_full_group_by` in sql_mode? 

Thanks

**Comment by bjendres** [comment]
Thanks @masetto. Yes, this module was developed when ``mysql`` was still shipped without ``only_full_group_by`` enabled. I recommend turning it off for CiviCRM, because you might encounter similar issues with other extensions as well.

That being said, we should -of course- fix that in one of the next versions.

**Comment by bjendres** [comment]
PRs always welcome :)

**Comment by masetto** [comment]
@bjendres please can you help me to creare a PR? I created a local branch `hotfix/sql_mode_only_full_group_by`, but I'm not have permissions to push it. My PR has a very minimal modification of a file.

**Comment by bjendres** [comment]
Hi @masetto, thanks for looking into this. If you have created this branch in a forked GitHub repository you should simply be able to create pull request. However, I didn't see a forked repository on your GitHub account. If you haven't forked the extension yet, this would be the place to start.

**Comment by masetto** [comment]
Hi @bjendres, I create my firstly PR, be patient if I didn't do everything right

**Comment by bjendres** [comment]
@masetto don't worry, it's all good :)

BTW: you should know that if it's too much work (or too risky) to fix the sql query, you can aways wrap the execution in ``CRM_Utils_SQL::disableFullGroupByMode()`` to make it work as-is.

**Comment by masetto** [comment]
@bjendres for this query (file ):
```
      SELECT
        civicrm_sdd_mandate.id                                  AS mandate_id,
        civicrm_contribution_recur.id                           AS rcur_id,
        civicrm_contribution_recur.start_date                   AS start_date,
        civicrm_contribution_recur.end_date                     AS end_date,
        civicrm_contribution_recur.next_sched_contribution_date AS next_collection_date,
        last.receive_date                                       AS last_collection_date,
        last.cancel_reason                                      AS last_cancel_reason,
        civicrm_sdd_mandate.status                              AS status,
        civicrm_sdd_mandate.reference                           AS reference,
        cancel_reason.note                                      AS cancel_reason,
        civicrm_financial_type.name                             AS financial_type,
        civicrm_campaign.title                                  AS campaign,
        civicrm_sdd_mandate.reference                           AS reference,
        civicrm_contribution_recur.frequency_interval           AS frequency_interval,
        civicrm_contribution_recur.frequency_unit               AS frequency_unit,
        civicrm_contribution_recur.currency                     AS currency,
        civicrm_contribution_recur.amount                       AS amount,
        IF(civicrm_sdd_mandate.status IN ('FRST', 'RCUR'), 'sepa-active', 'sepa-inactive')           
                                                                AS class
      FROM civicrm_sdd_mandate
      LEFT JOIN civicrm_contribution_recur ON civicrm_contribution_recur.id = civicrm_sdd_mandate.entity_id
      LEFT JOIN civicrm_financial_type     ON civicrm_financial_type.id = civicrm_contribution_recur.financial_type_id
      LEFT JOIN civicrm_campaign           ON civicrm_campaign.id = civicrm_contribution_recur.campaign_id
      LEFT JOIN civicrm_contribution last  ON last.receive_date = (SELECT MAX(receive_date) FROM civicrm_contribution
                                                                   WHERE contribution_recur_id = civicrm_contribution_recur.id
                                                                     AND contribution_status_id != 2)
      LEFT JOIN civicrm_note cancel_reason ON cancel_reason.entity_id = civicrm_contribution_recur.id
                                            AND cancel_reason.entity_table = 'civicrm_contribution_recur'
                                            AND cancel_reason.subject = 'cancel_reason'
      WHERE civicrm_sdd_mandate.contact_id = %1
        AND civicrm_sdd_mandate.type = 'RCUR'
        AND civicrm_sdd_mandate.entity_table = 'civicrm_contribution_recur'
      GROUP BY civicrm_sdd_mandate.id
      ORDER BY civicrm_contribution_recur.start_date DESC, civicrm_sdd_mandate.id DESC;";
```
there is no aggregate field. 
Do you expect each mandate to have more recurrent contributions? If so, maybe there is something wrong with the logic....

Do I add each field in the group by (makes no sense to me) or do I use the min function? and for which fields?

**Comment by bjendres** [comment]
As I said above, I wouldn't want to touch those as long as it works. The idea would be to wrap those in ``CRM_Utils_SQL::disableFullGroupByMode()`` directives like [HERE](https://github.com/systopia/de.systopia.donrec/blob/09bd2a65ae85b0f764087ccc1bfebae84d849062/CRM/Donrec/Logic/Receipt.php#L558-L560), so it doesn't cause any trouble. I'd be happy to accept PRs for that.

But If you really want to redesign all those queries, they need to be tested and reviewed well, and I don't think I currently have the time for that.

**Comment by eptbertram** [comment]
Disabling SQL_mode only_full_group_by solved my problem. A solution would be awesome.

**Comment by bjendres** [comment]
@masetto's changes have been merged to ``master`` and will be released with ``1.7``. Thanks @masetto
